Expanded the ConsulOp algebra to include a number of previously unsupported operations

aldiyen opened the pull request:
I'm guessing you're using Registrator or something along those lines to register your services. I need to register services from within my application, so I'm probably gonna add some stuff related to deregistering services and adding health checks in addition to this commit. Let me know if you have any preferences on design surrounding agent-based actions vs. more generic API calls, and of course I am happy to change formatting or JSON generation or whatever as you desire!

Personally I think it's probably OK to have everything in one place, especially since ConsulOp is sealed, though it's gonna get kinda long if every API call is implemented.

p.s. I'm using Sublime rather than an IDE, and your wildcard imports kill me a little bit every time I have to figure out where types are coming from. ;)

rossabaker commented on core/src/main/scala/HelmOp.scala:
```diff
@@ -21,6 +21,14 @@ object ConsulOp {
+    id:      Option[String],
+    tags:    Option[NonEmptyList[String]],
+    address: Option[String],
+    port:    Option[Int]
```
I wonder if we should support [all the parameters](https://www.consul.io/api/agent/service.html#parameters) supported by `AgentRegisterService` out of the gate to avoid a breaking change later when someone wants to specify one.

aldiyen replied:
I'm cool with adding the other service API calls. Specifically, it looks like that'd be:
- List Services
- Deregister Service
- Enable Maintenance Mode

Regarding adding the missing parameters to AgentRegisterService, sure. It looks like the only I haven't added here are the EnableTagOverride, Check, and Checks.

Welp, I'll work on that stuff now. Thanks!

rossabaker commented on core/src/main/scala/Interval.scala:
```diff
@@ -0,0 +1,36 @@
+  // seems like overkill.
+  // Also note that Consul supports intervals like "2h45m",
+  // seemingly because Go's time formatting library does,
+  // but that doesn't seem useful in a health check interval.
```
Yeah, that seems like a reasonable omission to me.

rossabaker commented on core/src/main/scala/Interval.scala:
```diff
@@ -0,0 +1,36 @@
+object Interval {
+
+  // So I considered using Scala's FiniteDuration type, but it
+  // seems like overkill.
```
My first instinct was to argue against this, but I guess it gives you a nice way to centralize the serialization into Go durations as more pop up in the future.

aldiyen replied:
I'm open to alternative implementations for this, if you have something else in mind. I think using the type system to enforce properly formed interval values is preferable, but beyond that I'm not married to any particular design.

rossabaker replied:
I don't have a better idea.  This works for me.

aldiyen commented:
I can bump the version, no problem. I also need some of the health checks API, so if you're concerned about changing the algebra do you want to wait for that stuff as well before triggering a new version?

rossabaker commented:
Yes, if you think you have more additions coming soon, let's wait for those and queue them up into one major release.  But I'm glad to see these additions.  Thank you!

aldiyen commented:
Hey okay, so I added all of the health APIs (though I only really needed one of them). In the process, I changed a few things and made some design decisions. I don't know if you folks have a Slack instance or some place to discuss things, but real-time chat seems to me to be a better venue for complex discussions than this... that is, assuming you have feedback other than "awesome this is great, thanks Matt!" ;)

timperrett commented:
@aldiyen thanks for doing this - looks good! We do indeed have a Gitter channel (There's a Verizon Open Source/Lobby channel).

Right now we're a little short-staffed because of the summer vacations etc, but i'll see what I can do about getting a review cycle on this by the end of the week. Does that sound acceptable?

Thanks!

aldiyen commented:
Sure, that sounds good. Thanks!

aldiyen commented:
I think the only operation name that conflicts with other parts of the API is `Delete`. My primary motivation was consistency -- once I realized I would have to prefix some operation names, it seemed weird to leave only a subset without such a prefix. That said, I can see your concern about backwards compatibility. I don't feel strongly about the prefix on those operation/function names.

aldiyen commented:
Yeah there are `Delete` operations in the ACL, Prepared Query, and Session APIs.
